docs/source/toolkit/testing.rst

```diff
@@ -0,0 +1,213 @@
+
+
+Mocks
+^^^^^
+
+The toolkit provides various `mocks
+<https://docs.python.org/3/library/unittest.mock.html#the-mock-class>`_ which
+can be used individually to stub out your Action Provider's authentication. You
+should use these directly if you are writing an Action Provider using a
+non-Flask framework or if you've decided not to use the built in Flask helpers.
+
+.. note::
+
+    This toolkit uses these mocks within the
+    :code:`globus_action_provider_tools.testing.fixtures` module.  
+
+
+.. _mock-authstate:
+
+Mock AuthState
+--------------
+
+An *AuthState* represents a requestor's authentication status and Globus Auth
+information. Every request should have its token validated via the
+*TokenChecker*'s :code:`check_token` method, which in turns generates an
+*AuthState* object.
+
+During testing, it is desirable to not need to provide valid tokens with every
+request. Use the :code:`mock_authstate` mock to generate a stubbed out
+*AuthState* object that won't validate requestor properties against Globus Auth.
+This is most useful when used in a patch as the return value for the
+*TokenChecker*'s :code:`check_token` method:
+
+.. code-block:: python
+
+    from unittest import mock     
+
+    import pytest
+    from globus_action_provider_tools.testing.mocks import mock_authstate
+
+    ...
+
+    @pytest.fixture
+    def client():
+        with mock.patch(
+            "globus_action_provider_tools.authentication.TokenChecker.check_token",
+            return_value=mock_authstate(),
+        ):
+            yield app.test_client()
+
+The example above creates a fixture which can be used to create a client that
+can make unauthenticated HTTP requests against an Action Provider. 
+
+
+Mock TokenChecker
+-----------------
+
+Because the *TokenChecker* is this toolkit's authentication workhorse, it's
+possible to entirely replace the the *TokenChecker* with a mock object. Doing so
+will allow your Action Provider to start up without validating its Client ID or
+Secret and will also allow unauthenticated requests to be made against it. This
+mock provides a simple way of completely removing your app's authentication
+during testing.
+
+
+.. code-block:: python
+
+    from unittest import mock
+
+    import pytest
+    from globus_action_provider_tools.testing.mocks import mock_tokenchecker
+
+    @pytest.fixture
+    def client():
+        with mock.patch(
+            "my_package.my_app.get_tokenchecker",
+            return_value=mock_tokenchecker(),
+        ):
+            app = create_app()
+            app.config["TESTING"] = True
+            yield app.test_client()
+
+
+.. note::
+
+    This example will only work if there's a function or method that is used to
+    create the TokenChecker instance. It demonstrates how you can patch a
+    function or a method to return the Mock TokenChecker. Internally, the Mock
+    TokenChecker will generate the :ref:`mock-authstate` objects described
```

Comment:
Ah, this kinda makes sense to have the other description first. Perhaps makes my previous comment not so good.